### PR TITLE
arch/risc-v/src/mpfs/mpfs_corespi.c: Add support for multiple bit widths

### PR DESCRIPTION
**Pull request by jlaitine**
The corespi fpga block supports just one frame length, which is defined when the block is instantiated on the FPGA.

This adds support for emulating different frame lengths if they are multiples of 8-bit. That is, with 8-bit corespi one can do 8,16 and 24-bit transfers.

This is implemented by simply writing several 8-bit frames for a single word when needed.

### CODE REVIEW

File: arch/risc-v/src/mpfs/mpfs_corespi.c [hunk 1]
```diff
@@ -720,6 +731,52 @@ static int mpfs_spi_sem_waitdone(struct mpfs_spi_priv_s *priv)
...
+ *   None
+ *
+ ****************************************************************************/
+
+static inline void mpfs_spi_write_tx8(struct mpfs_spi_priv_s *priv,
+                                      unsigned nframes,
+                                      bool last)
+{
+  uint8_t *data8 = (uint8_t *)priv->txbuf;
+  unsigned pos = priv->tx_pos++;
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:755]: What is this ++ supposed to do ? Aren't you supposed to move tx_pos forward by nframes * frame_size ?

File: arch/risc-v/src/mpfs/mpfs_corespi.c [hunk 2]
```diff
@@ -720,6 +731,52 @@ static int mpfs_spi_sem_waitdone(struct mpfs_spi_priv_s *priv)
...
+
+  /* Calculate octet position in the tx-buffer */
+
+  pos *= nframes;
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:759]: Wat

File: arch/risc-v/src/mpfs/mpfs_corespi.c [hunk 3]
```diff
@@ -720,6 +731,52 @@ static int mpfs_spi_sem_waitdone(struct mpfs_spi_priv_s *priv)
...
+
+  /* Write all but last 8-bit data */
+
+  while (--nframes > 0)
+    {
+      putreg32((uint32_t)data8[pos + nframes], MPFS_SPI_TX_DATA);
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:765]: Writes the data in reverse, or am I missing something ?

File: arch/risc-v/src/mpfs/mpfs_corespi.c (hunk 1 above)
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:755]: Maybe this ++ can be simplified to nframes * frame_size ?

File: arch/risc-v/src/mpfs/mpfs_corespi.c [hunk 4]
```diff
@@ -781,6 +837,39 @@ static void mpfs_spi_load_tx_fifo(struct mpfs_spi_priv_s *priv,
...
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static inline void mpfs_spi_read_rx8(struct mpfs_spi_priv_s *priv,
+                                     unsigned nframes)
+{
+  uint8_t *data8 = (uint8_t *)priv->rxbuf;
+  unsigned pos = priv->rx_pos++;
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:859]: Ditto

**Review by pussuw — APPROVED** [review]
LGTM, very confusing / hard to follow logic but seems to work

File: arch/risc-v/src/mpfs/mpfs_corespi.c (hunk 3 above)
Comment (jlaitine) [reply arch/risc-v/src/mpfs/mpfs_corespi.c:765]: Yes, the SPI frame is sent msb bit first, so making a 16-bit transfer out of 2 8-bit transfers requires sending the msb byte first, lsb byte last.

File: arch/risc-v/src/mpfs/mpfs_corespi.c (hunk 1 above)
Comment (jlaitine) [reply arch/risc-v/src/mpfs/mpfs_corespi.c:755]: nope, the priv->tx_pos counts the SPI words (priv->nbits size words), and this function just writes one of those. E.g. if priv->nbits == 16, but we have 8-bit corespi, this means that nframes == 2 (need to combine 2 8-bit spi frames to write one requested 16-bit frame one - this function writes 2 8-bit spi frames).

Comment (jlaitine) [reply arch/risc-v/src/mpfs/mpfs_corespi.c:755]: Oh, or did you mean to just put this on one line? 

unsigned pos = nframes * priv->tx_pos++;

I felt that it makes it even more confusing....

Comment (pussuw) [reply arch/risc-v/src/mpfs/mpfs_corespi.c:755]: No, this is fine as is

Comment (jlaitine) [reply arch/risc-v/src/mpfs/mpfs_corespi.c:755]: or, could keep the incrementing of priv->tx_pos in "mpfs_spi_load_tx_fifo", and just pass the current word index as an argument....
